esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6118,7 +6130,329 @@ bool CWsDfuEx::onDFUFileAccess(IEspContext &context, IEspDFUFileAccessRequest &r
             userDesc.setown(createUserDescriptor());
             userDesc->set(userID.str(), context.queryPassword(), context.querySignature());
         }
-        getFileAccess(context, userDesc, accessType, req, resp);
+        getFileAccess(context, userDesc, accessType, requestBase, resp.updateAccessInfo());
+    }
+    catch (IException *e)
+    {
+        FORWARDEXCEPTION(context, e,  ECLWATCH_INTERNAL_ERROR);
+    }
+    return true;
+}
+
+bool CWsDfuEx::getGroupNameByClusterNameAndType(IPropertyTree *envTree, const char *clusterName, ClusterType clusterType, StringBuffer &groupName)
+{
+    StringBuffer xpath;
+    if (clusterType == ThorLCRCluster)
+        xpath.set("Software/ThorCluster");
+    else if (clusterType == RoxieCluster)
+        xpath.set("Software/RoxieCluster");
+    else
+        xpath.set("Software/EclAgentProcess");
+
+    Owned<IPropertyTreeIterator> it= envTree->getElements(xpath.str());
+    ForEach(*it)
+    {
+        IPropertyTree &cluster = it->query();
+        if (strieq(clusterName, cluster.queryProp("@name")))
+        {
+            if (clusterType != HThorCluster)
+                getClusterGroupName(cluster, groupName);
+            else
+                groupName.setf("hthor__%s", clusterName);
+            return true;
+        }
+    }
+    return false;
+}
+
+StringBuffer &CWsDfuEx::getGroupNameByClusterName(const char *clusterName, StringBuffer &groupName)
```

Comment:
Normal convention (certainly not universal) is to have variables that are output (i.e. groupName here) as the leading parameters.